### DISCUSSION TITLE
SMT implementation on gem5 in SE mode

### DISCUSSION THREAD

**Discussion by ishitachaturvedi** [gem5/gem5#282] [gem5-dev]
I am trying to implement SMT on gem5. Gem5 currently supports pthreads in SE mode, where each core runs one thread.

I did the following tests:

1. Ran 2 different workloads on 2 SMT threads (on the same core) in gem5.
The run command I used was:
`build/X86/gem5.opt se_SMT.py --cmd=test_codes/single_thread --caches --l1d_size=32kB --l1d_assoc=8 --l1i_size=32kB --l1i_assoc=8 --l2cache --cpu-type=DerivO3CPU --mem-type=DDR4_2400_16x4 --mem-size=64GB --mem-channels=2 --num-cpus=7 --smt`

To attach 2 workloads to 2 threads I did:
```
for wrkld in workloads:
        if args.smt:
            for i in range(numThreads):
                process = Process(pid=100 + idx)
                process.executable = wrkld
                process.cwd = os.getcwd()
                process.gid = os.getgid()

                if args.env:
                    with open(args.env, "r") as f:
                        process.env = [line.rstrip() for line in f]

                if len(pargs) > idx:
                    process.cmd = [wrkld] + pargs[idx].split()
                else:
                    process.cmd = [wrkld]

                if len(inputs) > idx:
                    process.input = inputs[idx]
                if len(outputs) > idx:
                    process.output = outputs[idx]
                if len(errouts) > idx:
                    process.errout = errouts[idx]

                multiprocesses.append(process)
                idx += 1
```
This test ran fine, and the output was expected.

2. Ran one workload on 2 SMT threads on gem5. 
The run command I used was: 
`build/X86/gem5.opt se_2_SMT_1_workload.py --cmd=test_codes/hello_pthreads --caches --l1d_size=32kB --l1d_assoc=8 --l1i_size=32kB --l1i_assoc=8 --l2cache --cpu-type=DerivO3CPU --mem-type=DDR4_2400_16x4 --mem-size=64GB --mem-channels=2 --num-cpus=7 --smt`

To give 2 SMT threads the same workload I did:
``` 
if args.smt:
            for i in range(numThreads):
                process = Process(pid=100 + idx)
                process.executable = wrkld
                process.cwd = os.getcwd()
                process.gid = os.getgid()

                if args.env:
                    with open(args.env, "r") as f:
                        process.env = [line.rstrip() for line in f]

                if len(pargs) > idx:
                    process.cmd = [wrkld] + pargs[idx].split()
                else:
                    process.cmd = [wrkld]

                if len(inputs) > idx:
                    process.input = inputs[idx]
                if len(outputs) > idx:
                    process.output = outputs[idx]
                if len(errouts) > idx:
                    process.errout = errouts[idx]

                multiprocesses.append(process)
```
On running this, I got the **error: build/X86/sim/process.cc:142: fatal: fatal condition !ret_pair.second occurred: _pid 100 is already used**

3. To run pthreads workload on multicore (with 1 thread per core) used the command:
`build/X86/gem5.opt se_1_workload.py --cmd=test_codes/hello_pthreads --caches --l1d_size=32kB --l1d_assoc=8 --l1i_size=32kB --l1i_assoc=8 --l2cache --cpu-type=DerivO3CPU --mem-type=DDR4_2400_16x4 --mem-size=64GB --mem-channels=2 --num-cpus=7 --smt`

While this run had no errors, I think pthers_join is ignored, because the output looks like:
```
Hello, World! from thread #0
All threads have completed.
Hello, World! from thread #1
```

4. Ran 2 pthreads workloads on 2 cores such that:
Core0:
SMT0: Workload A
SMT1: Workload B
Core1:
SMT0: Workload A
SMT1: Workload B

The run command I used was: 
`build/X86/gem5.opt se_SMT.py --cmd=test_codes/hello_pthreads --caches --l1d_size=32kB --l1d_assoc=8 --l1i_size=32kB --l1i_assoc=8 --l2cache --cpu-type=DerivO3CPU --mem-type=DDR4_2400_16x4 --mem-size=64GB --mem-channels=2 --num-cpus=7 --smt`

I made the following change to the config file to do this:
```
for wrkld in workloads:
        if args.smt:
            for i in range(numThreads):
                process = Process(pid=100 + idx)
                process.executable = wrkld
                process.cwd = os.getcwd()
                process.gid = os.getgid()

                if args.env:
                    with open(args.env, "r") as f:
                        process.env = [line.rstrip() for line in f]

                if len(pargs) > idx:
                    process.cmd = [wrkld] + pargs[idx].split()
                else:
                    process.cmd = [wrkld]

                if len(inputs) > idx:
                    process.input = inputs[idx]
                if len(outputs) > idx:
                    process.output = outputs[idx]
                if len(errouts) > idx:
                    process.errout = errouts[idx]

                multiprocesses.append(process)
                idx += 1
```
      

This configuration did not support pthread_exit and pthread_join. When these are commented out all the print statements are displayed but the simulation gets stuck after the print statements.
      
I want to fix case 2 and have 1 workload be processed by 2 SMT threads on 1 core and case 4, for pthreads processes not to get stuck. 

I am attaching the test files and all the config files. Please let me know how I can fix this. Thanks! @powerjg 
[Pthreads_error.zip](https://github.com/gem5/gem5/files/12551533/Pthreads_error.zip)

**Comment by BobbyRBruce** [comment]
This is really more of an issue, or potential issue, so I opened up one here: #338. SMT should work in gem5 SE mode.

I'm going to close this discussion and try to replicate the issue as highlighted in #338, so let's continue discussion there.
